src/main/java/org/commonjava/util/partyline/FileTree.java

```diff
@@ -327,46 +327,55 @@ boolean tryLock( File file, String ownerName, String label, LockLevel lockLevel,
         while ( end < 1 || System.currentTimeMillis() < end )
         {
             T result = tryLock( realFile, Thread.currentThread().getName(), "Open File for " + ( doOutput ? "output" : "input" ),
-                                           doOutput ? LockLevel.write : read, WAIT_TIMEOUT, unit, ( opLock ) -> {
+                                           doOutput ? LockLevel.write : read, timeout, unit, ( opLock ) -> {
                         FileEntry entry = entryMap.get( realFile.getAbsolutePath() );
-                        if ( entry.file != null )
+                        boolean proceed = false;
+                        synchronized ( entry )
```

Comment:
IMO, "the only time GC can remove a lock is if the thread that used the string and the lock is done" isn't good enough to prevent a collision. One can break this by threading off the operation so the withOpLock returns immediately and GC is free to collect the path/lock entry. Meanwhile, the back end thread still r/w the file with a detached lock object. I am not sure this use case is valid, but anyway depending on GC behavior is fragile. If the original intention is to "a. ensure one file/path only has one oplock, b. avoid map being too big", probably we can tackle it via a robust cache (something like apache SoftHashMap, etc). Or embedding the op lock inside the FileEntry (so the file and the lock bound together, treated as a whole). I am not sure I understand partyline in depth. Just my thoughts.